### DISCUSSION TITLE
Libbi Temperature

### DISCUSSION THREAD

**Discussion by MalcolmSpencer** [CJNE/ha-myenergi#526] [Ideas]
I have an HA automation that calculates my target Libbi SOC based on the Solcast forecast and then calculates the required charge start time so that the target SOC is reached at 5am when my cheap rate finishes. I hope to integrate this to the Libbi when @trizmark completes the integration being worked on. 

The problem is that the charge time varies significantly with temperature (especially below 10 degrees C) so it would be very helpful to know the Libbi temperature. Temperature isn't on the status API https://s18.myenergi.net/cgi-jstatus-Lxxxxx but the historic API https://s18.myenergi.net/cgi-jday-Lxxxxx-yyyy-mm-dd has 

ivt - Inverter temp
batt	- Battery temp
batrt - Battery radiator temp
ambt - Ambient temp
contt - Controller temp

If temperature was available then this could be used to optimise the charge time. It would be great if some or all the temperatures could be available on the integration.

**Comment by MalcolmSpencer** [comment]
In the Libbi integration, there is a state 172: "BMS Charge Temperature Low"
I presume it maps to the KStar 'W22 Bms Chg Temp Low alarm' but I can't find any info on what temperature this is actually triggered.
Its also a warning so I presume that it affects the charge rate rather than stops the charge.

Has anyone ever seen this status?
I had some low temperatures and slow charging around the 12th Dec 2023 but I didn't have HA at that time so I can't check my Libbi status.

If this does indicate a slow charge situation that would be really helpful.

Thanks for any help

**Comment by trizmark** [comment]
I have seen that status. IIRC the battery temperature was something like 2 or 3C. Charging was severely limited; it was 940W on the 5kW inverter.
That status only came on though when temperatures were low for a fair number of consecutive days.

**Comment by MalcolmSpencer** [comment]
I have noticed that the battery is taking longer to charge in this colder weather. Looking at the data from the historic API it seems that the charge rate is throttled back depending on the battery temperature. 

![image](https://github.com/user-attachments/assets/0c4334f3-66a9-4f72-88a6-de2a4eba4ae9)


However I haven't seen a state 172: "BMS Charge Temperature Low" 
With these lower charge rates, I can't charge to 100% in the 3hours cheap elec I have from Octopus flux so I am thinking I need to do a sort of pre-charge starting 30 mins before the cheap period to warm the battery so it is charging at a higher rate during the cheap period.

I would be interested to hear if other people have encountered this issue and how you solved it.

@trizmark

**Reply by G6EJD** [reply]
This is a well known issue with the inferior Libbi product where the BMS is reducing the charge rate; unlike all other battery systems, but ME are reputably going to issue a firmware update this coming week to help reduce the issue. Apparently the battery heater is insufficient to warm the battery mass and preempt the problem. It's what happens to early adopters of a poor product.
BTW GitHub is not the forum to discuss hardware issues and it's nothing to do with this integration!

**Reply by trizmark** [reply]
@G6EJD If you have nothing constructive add, then please find another hobby. Thank you.

@MalcolmSpencer I'm in the same boat with Octopus Flux. 3 hour charging slot is not enough for winter. I have tapped into the Modbus interface, so I have some real nifty graphs about all the internal workings of the libbi. The issue is that the heater only activates once the grid charging starts and takes about an hour to bring the packs up to temperature. (I have not added the insulation back yet, but I should!)
I've submitted a feature request to start the pre-heating at least 30 minutes before the actual start of the grid charging. Let's see. 
Which SW version are you on? I've got a non-released beta from ME engineering at the beginning of the year.

You can see the preheat in the graph below. It starts on the dot at 2am (and then again once the sun is out). IMO the grid charging should start at least 30 mins before grid charging is activated.

![image](https://github.com/user-attachments/assets/7e54631f-175a-43bc-a17c-c5aa9e66e1d2)

**Comment by HLFCode** [comment]
I'm guessing this is only an issue for units with one battery.
Since multiple batteries are in parallel and share the charge power 2 bats will only see max 2.5kW each.
I certainly don't notice any throttling with my 4 bat pack even during the recent subzero weather.

Didn't know they have a heater!

I would like to know the battery temperatures though :-)

**Reply by trizmark** [reply]
Nah, I got 4 batteries, so it's an issue with all of them.
Here's a graph for the CT clamp that's on the libbi feed to the house. As you can see the initial charge rate is limited to 1.1kW while the batteries are heating up. After an hour the charge rate ramps up and within 20 mins after that we're at full 5kW charging rate.
I am still looking if it would be possible to insert that temperature data into HA - if nothing else I can extend pymyenergy and you can save it from there.

![image](https://github.com/user-attachments/assets/a7dd620c-5f74-43a4-9dc6-73ca7a19f81c)

**Comment by HLFCode** [comment]
Strange
It was vvv cold last week and mine slapped straight to 4.5kW at the 23:23 tariff change
![image](https://github.com/user-attachments/assets/0342ce9a-0bb1-4664-8b51-d8131d30b766)
Batteries in an unheated barn so must have got down to near zero

**Comment by MalcolmSpencer** [comment]
@HLFCode it looks like your batteries were fairly active. I only have 10Kw batteries and very often they run out around 10pm and then sit there until 2am waiting to charge. 

I wonder if it would be better to set up an automation to keep charging and allow them to discharge during cold weather. i.e. If Soc <3% and battery temp <9 degrees then charge to 20%.

Would this be better than letting them get cold and then pre-heating?
Would this cost more than pre-heat?

To do this we would need to pull the battery temp from the Myenergi historic API say every 5 to 10 mins.

I am on firmware build  3702S5.437
